---
title: Correcting batch effects in single-cell RNA-seq data
author: Aaron Lun
date: "2018-05-17"
output: 
  BiocStyle::html_document:
    fit_caption: false
---



# Introduction

Large scRNA-seq datasets are often generated in multiple batches due to logistical constraints.
This results in batch effects due to uncontrollable differences in the processing of different batches, e.g., changes in operator, differences in reagent quality.
Batch effects are problematic as they can be major drivers of heterogeneity in the data, masking the relevant biological differences and complicating interpretation of the results.

Here, we will explore a few methods for removing these batch effects.
We will be using a haematopoiesis dataset, generated by two groups with different technologies and involving different parts of the haematopoietic lineage.
The [Nestorowa](https://doi.org/10.1182/blood-2016-05-716480) dataset uses Smart-seq2 to profile haematopoietic stem cell and precursor populations.
In contrast, the [Paul](https://doi.org/10.1016/j.cell.2015.11.013) dataset uses MARS-seq to profile the myeloid progenitors. 

For the sake of speed, we've already processed the two datasets to call highly variable genes - see `prepare.html` for details.
We load in the various R objects using the `load()` function.


```r
library(SingleCellExperiment)
load("objects.Rda")
ls()
```

```
## [1] "decN" "decP" "fitN" "fitP" "sceN" "sceP"
```

<div class="alert alert-warning">
**Exercise:** 

We inspect some of these objects:


```r
sceN
```

```
## class: SingleCellExperiment 
## dim: 46170 1920 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(46170): ENSMUSG00000000001 ENSMUSG00000000003 ...
##   ERCC-00170 ERCC-00171
## rowData names(0):
## colnames(1920): HSPC_007 HSPC_013 ... Prog_852 Prog_810
## colData names(1): CellType
## reducedDimNames(0):
## spikeNames(1): ERCC
```


```r
sceP
```

```
## class: SingleCellExperiment 
## dim: 18092 1862 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(18092): ENSMUSG00000007777 ENSMUSG00000024442 ...
##   ENSMUSG00000090626 ENSMUSG00000032324
## rowData names(9): is_feature_control mean_counts ... n_cells_counts
##   pct_dropout_counts
## colnames(1862): W75265 W38465 ... W37729 W37745
## colData names(15): is_cell_control total_features_by_counts ...
##   pct_counts_top_200_features pct_counts_top_500_features
## reducedDimNames(0):
## spikeNames(0):
```


```r
decN
```

```
## DataFrame with 46170 rows and 6 columns
##                                   mean              total
##                              <numeric>          <numeric>
## ENSMUSG00000000001    6.34334751550413   7.93088831876345
## ENSMUSG00000000003 0.00979007329151634 0.0253290312805928
## ENSMUSG00000000028     3.9902456058454   9.35478320944155
## ENSMUSG00000000031   0.111949614544398  0.460678594211501
## ENSMUSG00000000037    1.08606588143884   4.13900720759496
## ...                                ...                ...
## ERCC-00164          0.0713966829808465  0.398275307021139
## ERCC-00165            5.19789272169338   9.98136111063097
## ERCC-00168           0.102595858457219   0.57913449381599
## ERCC-00170            2.12765568078709   7.76421496013614
## ERCC-00171            12.4971928247349 0.0723121633598244
##                                    bio               tech
##                              <numeric>          <numeric>
## ENSMUSG00000000001      1.211919636143   6.71896868262045
## ENSMUSG00000000003 -0.0268192853388259 0.0521483166194187
## ENSMUSG00000000028  -0.577730972579486   9.93251418202104
## ENSMUSG00000000031  -0.132253854687009  0.592932448898509
## ENSMUSG00000000037  -0.815115814425082   4.95412302202004
## ...                                ...                ...
## ERCC-00164          0.0179700137519021  0.380305293269237
## ERCC-00165            1.63544143039118   8.34591968023979
## ERCC-00168          0.0326421136331594   0.54649238018283
## ERCC-00170          -0.203730241449941   7.96794520158608
## ERCC-00171         -0.0563575120797276  0.128669675439552
##                                 p.value                  FDR
##                               <numeric>            <numeric>
## ENSMUSG00000000001 6.00580100025002e-08 4.75490203590242e-06
## ENSMUSG00000000003                    1                    1
## ENSMUSG00000000028    0.966156176729986                    1
## ENSMUSG00000000031    0.999999999999966                    1
## ENSMUSG00000000037    0.999999965828181                    1
## ...                                 ...                  ...
## ERCC-00164                           NA                   NA
## ERCC-00165                           NA                   NA
## ERCC-00168                           NA                   NA
## ERCC-00170                           NA                   NA
## ERCC-00171                           NA                   NA
```


```r
decP
```

```
## DataFrame with 18092 rows and 6 columns
##                                  mean              total
##                             <numeric>          <numeric>
## ENSMUSG00000007777 0.0371305230862186 0.0400754051582754
## ENSMUSG00000024442  0.150766095383802  0.174716159359473
## ENSMUSG00000078880                  0                  0
## ENSMUSG00000093989  0.956864963377254  0.918615249367187
## ENSMUSG00000107002   0.48964365559762  0.509101611842776
## ...                               ...                ...
## ENSMUSG00000028226                  0                  0
## ENSMUSG00000000085   0.13057491278379  0.151673880538042
## ENSMUSG00000038178 0.0734146377039899 0.0933064410612593
## ENSMUSG00000090626 0.0321911448581939 0.0391420172945157
## ENSMUSG00000032324    0.1426629746993  0.189926881940395
##                                     bio               tech
##                               <numeric>          <numeric>
## ENSMUSG00000007777 -0.00983619395322893 0.0499115991115043
## ENSMUSG00000024442  -0.0121316752377102  0.186847834597183
## ENSMUSG00000078880                    0                  0
## ENSMUSG00000093989  -0.0115400651780048  0.930155314545192
## ENSMUSG00000107002  -0.0420411886549673  0.551142800497743
## ...                                 ...                ...
## ENSMUSG00000028226                    0                  0
## ENSMUSG00000000085 -0.00875394964170809   0.16042783017975
## ENSMUSG00000038178 -0.00537899780232341 0.0986854388635827
## ENSMUSG00000090626 -0.00412996983509422 0.0432719871296099
## ENSMUSG00000032324   0.0119188516711214  0.178008030269273
##                               p.value               FDR
##                             <numeric>         <numeric>
## ENSMUSG00000007777  0.999999999943339                 1
## ENSMUSG00000024442  0.978003138518491                 1
## ENSMUSG00000078880                  1                 1
## ENSMUSG00000093989  0.643969515694305                 1
## ENSMUSG00000107002    0.9912918542818                 1
## ...                               ...               ...
## ENSMUSG00000028226                  1                 1
## ENSMUSG00000000085   0.95398580903467                 1
## ENSMUSG00000038178  0.953802833477046                 1
## ENSMUSG00000090626  0.998643468362498                 1
## ENSMUSG00000032324 0.0222451626163784 0.156416432979214
```
</div>

# Merging a set of highly variable genes

First, we need to obtain a common set of genes based on the Ensembl IDs.
This is why Ensembl IDs are so useful - they are, by definition, constant across experiments.


```r
common <- intersect(rownames(decN), rownames(decP))
length(common)
```

```
## [1] 17924
```

We combine the statistics using `combineVar` from *[scran](http://bioconductor.org/packages/scran)*.
This uses Fisher's method to combine p-values across batches.
We take all genes that are significant at a FDR of 5%.


```r
library(scran)
combined <- combineVar(decP[common,], decN[common,])
to.use <- common[combined$FDR <= 0.05]
length(to.use)
```

```
## [1] 2713
```

<div class="alert alert-warning">
**Exercise:** 


```r
# What happens if we forget to use [common,]?
try(combined2 <- combineVar(decP, decN))
```
</div>

# Removing the batch effect

## With a linear model

The simplest approach is to remove batch effects with a linear model, e.g., using `removeBatchEffect`.
In this case, we're basically shifting the expression values so that the mean is the same between batches for every gene.
Let's try doing that:


```r
library(limma)
batch <- rep(c("N", "P"), c(ncol(sceN), ncol(sceP)))
together <- cbind(logcounts(sceN)[to.use,], logcounts(sceP)[to.use,])
corrected <- removeBatchEffect(together, batch=batch)
```

We set up a `SingleCellExperiment` object again:


```r
sceT <- SingleCellExperiment(list(exprs=corrected))
sceT$Batch <- batch
```

...and we make a PCA plot to look at, using *[irlba](https://CRAN.R-project.org/package=irlba)* to get the first 50 PCs.
We specify `ntop=Inf` to tell `runPCA` to use all available genes in `to.use`.


```r
library(scater)
sceT <- runPCA(sceT, exprs_values="exprs", method="irlba",
    ncomponents=50, ntop=Inf)
plotPCA(sceT, colour_by="Batch")
```

<img src="answers_files/figure-html/unnamed-chunk-11-1.png" width="100%" />

Trying out a _t_-SNE plot:


```r
sceT <- runTSNE(sceT, use_dimred="PCA", perplexity=80)
plotTSNE(sceT, colour_by="Batch")
```

<img src="answers_files/figure-html/unnamed-chunk-12-1.png" width="100%" />

Even after `removeBatchEffect`, a strong batch effect still remains, for two reasons:

- differences in variance between UMI and read counts
- differences in cell type composition between data sets

<div class="alert alert-warning">
**Exercise:** 


```r
# How do I check what reduced dimension results are available?
reducedDimNames(sceT)
```

```
## [1] "PCA"  "TSNE"
```

We can get rid of the first effect to some extent using cosine normalization, but the second effect can't be resolved by _a priori_-defined linear models:


```r
together2 <- scran:::cosine.norm(together)
rownames(together2) <- to.use
corrected2 <- removeBatchEffect(together2, batch=batch)

# Same as before...
sceT2 <- SingleCellExperiment(list(exprs=corrected2))
sceT2$Batch <- batch
sceT2 <- runPCA(sceT2, exprs_values="exprs", method="irlba",
    ncomponents=50, ntop=Inf)
plotPCA(sceT2, colour_by="Batch")
```

<img src="answers_files/figure-html/unnamed-chunk-14-1.png" width="100%" />
</div>

## With MNN correction

A more sophisticated approach involves using the "mutual nearest neighbors" correction method (see https://www.nature.com/articles/nbt.4091/figures/1).
This involves finding pairs of cells - one per batch - where one cell in one batch is the nearest neighbour of a cell in the other batch, **and vice versa**.
Such pairs are likely to represent cells of the same type or state; we can use this to compute correction vectors between the two batches.


```r
corrected <- mnnCorrect(counts(sceN)[to.use,], counts(sceP)[to.use,], sigma=0.05)
```

We can combine the corrected matrices into a single object:


```r
out <- do.call(cbind, corrected$corrected)
colnames(out) <- NULL
sceM <- SingleCellExperiment(list(exprs=out),
    colData=DataFrame(Batch=batch))
```

... and we run a PCA using *[irlba](https://CRAN.R-project.org/package=irlba)* to get the first 50 PCs.


```r
sceM <- runPCA(sceM, exprs_values="exprs", method="irlba",
    ncomponents=50, ntop=Inf)
```

Plotting it indicates that the batch effect has been removed from the first two PCs.
In particular, the Paul dataset has been integrated into the Nestorowa dataset:


```r
plotPCA(sceM, colour_by="Batch")
```

<img src="answers_files/figure-html/unnamed-chunk-18-1.png" width="100%" />

The corrected values can be used for clustering, but it not for DE as the values are not interpretable as log-counts.
We suggest using the learnt structure (clusters) with the batch of origin as a blocking factor in the design matrix.

<div class="alert alert-warning">
**Exercise:** 


```r
# The _t_-SNE is a bit messier, but it shows the point:
sceM <- runTSNE(sceM, use_dimred="PCA", perplexity=80)
plotTSNE(sceM, colour_by="Batch")
```

<img src="answers_files/figure-html/unnamed-chunk-19-1.png" width="100%" />
</div>

# Session information


```r
sessionInfo()
```

```
## R version 3.5.0 Patched (2018-04-30 r74681)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.4 LTS
## 
## Matrix products: default
## BLAS: /home/cri.camres.org/lun01/Software/R/R-3-5-branch-release/lib/libRblas.so
## LAPACK: /home/cri.camres.org/lun01/Software/R/R-3-5-branch-release/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] scater_1.8.0                ggplot2_2.2.1              
##  [3] limma_3.36.1                scran_1.8.1                
##  [5] SingleCellExperiment_1.2.0  SummarizedExperiment_1.10.1
##  [7] DelayedArray_0.6.0          BiocParallel_1.14.1        
##  [9] matrixStats_0.53.1          Biobase_2.40.0             
## [11] GenomicRanges_1.32.2        GenomeInfoDb_1.16.0        
## [13] IRanges_2.14.9              S4Vectors_0.18.1           
## [15] BiocGenerics_0.26.0         knitr_1.20                 
## [17] BiocStyle_2.8.0            
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.5.1            dynamicTreeCut_1.63-1   
##  [3] edgeR_3.22.1             viridisLite_0.3.0       
##  [5] DelayedMatrixStats_1.2.0 shiny_1.0.5             
##  [7] assertthat_0.2.0         statmod_1.4.30          
##  [9] vipor_0.4.5              GenomeInfoDbData_1.1.0  
## [11] yaml_2.1.19              pillar_1.2.2            
## [13] backports_1.1.2          lattice_0.20-35         
## [15] glue_1.2.0               digest_0.6.15           
## [17] promises_1.0.1           XVector_0.20.0          
## [19] colorspace_1.3-2         cowplot_0.9.2           
## [21] htmltools_0.3.6          httpuv_1.4.3            
## [23] Matrix_1.2-14            plyr_1.8.4              
## [25] pkgconfig_2.0.1          bookdown_0.7            
## [27] zlibbioc_1.26.0          xtable_1.8-2            
## [29] scales_0.5.0             Rtsne_0.13              
## [31] later_0.7.2              tibble_1.4.2            
## [33] DT_0.4                   lazyeval_0.2.1          
## [35] magrittr_1.5             mime_0.5                
## [37] evaluate_0.10.1          beeswarm_0.2.3          
## [39] FNN_1.1                  shinydashboard_0.7.0    
## [41] tools_3.5.0              data.table_1.11.2       
## [43] stringr_1.3.1            Rhdf5lib_1.2.0          
## [45] munsell_0.4.3            locfit_1.5-9.1          
## [47] irlba_2.3.2              bindrcpp_0.2.2          
## [49] compiler_3.5.0           rlang_0.2.0             
## [51] rhdf5_2.24.0             grid_3.5.0              
## [53] RCurl_1.95-4.10          tximport_1.8.0          
## [55] rjson_0.2.18             htmlwidgets_1.2         
## [57] igraph_1.2.1             labeling_0.3            
## [59] bitops_1.0-6             rmarkdown_1.9           
## [61] gtable_0.2.0             reshape2_1.4.3          
## [63] R6_2.2.2                 gridExtra_2.3           
## [65] dplyr_0.7.4              bindr_0.1.1             
## [67] rprojroot_1.3-2          stringi_1.2.2           
## [69] ggbeeswarm_0.6.0         Rcpp_0.12.16            
## [71] xfun_0.1
```
